pkg/controller/awsprivatelink/awsprivatelink_controller.go

```diff
@@ -0,0 +1,1227 @@
+		curr.Status.Conditions,
+		hivev1.AWSPrivateLinkFailedClusterDeploymentCondition,
+		corev1.ConditionTrue,
+		reason,
+		err.Error(),
+		controllerutils.UpdateConditionIfReasonOrMessageChange)
+	conditions, readyChanged := controllerutils.SetClusterDeploymentConditionWithChangeCheck(
+		conditions,
+		hivev1.AWSPrivateLinkReadyClusterDeploymentCondition,
+		corev1.ConditionFalse,
+		reason,
+		err.Error(),
+		controllerutils.UpdateConditionIfReasonOrMessageChange)
+	if !readyChanged && !failedChanged {
+		return nil
+	}
+	curr.Status.Conditions = conditions
+	logger.Debug("setting AWSPrivateLinkFailedClusterDeploymentCondition to true")
+	return r.Status().Update(context.TODO(), curr)
+}
+
+func (r *ReconcileAWSPrivateLink) setProgressCondition(cd *hivev1.ClusterDeployment,
+	completed corev1.ConditionStatus,
+	reason string, message string,
+	logger log.FieldLogger) error {
+
+	curr := &hivev1.ClusterDeployment{}
+	errGet := r.Get(context.TODO(), types.NamespacedName{Namespace: cd.Namespace, Name: cd.Name}, curr)
+	if errGet != nil {
+		return errGet
+	}
+
+	conditions := curr.Status.Conditions
+
+	var failedChanged bool
+	if completed == corev1.ConditionTrue {
+		conditions, failedChanged = controllerutils.SetClusterDeploymentConditionWithChangeCheck(
+			conditions,
+			hivev1.AWSPrivateLinkFailedClusterDeploymentCondition,
+			corev1.ConditionFalse,
+			reason,
+			message,
+			controllerutils.UpdateConditionIfReasonOrMessageChange)
+	}
+
+	var readyChanged bool
+	if ready := controllerutils.FindClusterDeploymentCondition(conditions, hivev1.AWSPrivateLinkReadyClusterDeploymentCondition); ready == nil ||
+		ready.Status != corev1.ConditionTrue {
+		conditions, readyChanged = controllerutils.SetClusterDeploymentConditionWithChangeCheck(
+			conditions,
+			hivev1.AWSPrivateLinkReadyClusterDeploymentCondition,
+			completed,
+			reason,
+			message,
+			controllerutils.UpdateConditionIfReasonOrMessageChange)
+	}
+	if !readyChanged && !failedChanged {
+		return nil
+	}
+	curr.Status.Conditions = conditions
+	logger.Debug("setting AWSPrivateLinkReadyClusterDeploymentCondition to %s", completed)
+	return r.Status().Update(context.TODO(), curr)
+}
+
+func (r *ReconcileAWSPrivateLink) reconcilePrivateLink(cd *hivev1.ClusterDeployment, clusterMetadata *hivev1.ClusterMetadata, logger log.FieldLogger) (reconcile.Result, error) {
+	logger.Debug("reconciling PrivateLink resources")
+	awsClient, err := newAWSClient(r,
+		corev1.SecretReference{Name: cd.Spec.Platform.AWS.CredentialsSecretRef.Name, Namespace: cd.Namespace},
+		cd.Spec.Platform.AWS.Region)
+	if err != nil {
+		logger.WithError(err).Error("error creating AWS client for the cluster")
+		return reconcile.Result{}, err
+	}
+
+	if err := r.setProgressCondition(cd, corev1.ConditionFalse,
+		"DiscoveringNLB",
+		"discovering NLB for the cluster",
+		logger); err != nil {
+		logger.WithError(err).Error("failed to update condition on cluster deployment")
+		return reconcile.Result{}, err
+	}
```

Comment:
One thing I'm still worried about here, if there's a problem mid-way and we end up re-reconciling, there are going to be a lot of writes to ClusterDeployment status. We'll fail, set the failed condition, and then requeue, which will move the progress condition back through each state it already completed. It still feels like we need to track the state we're in here, or perhaps just monitor "did we do anything" a little better, and then only update the condition if we found something has to be done. In this case that would mean we drop the above set progress, and in the block below, if we can't find the NLB, we set WaitingForNLB reason.